File: java/LongestIncreasingSubsequence.java
import java.util.LinkedList;
import java.util.List;

public class LongestIncreasingSubsequence {
    public static List<Integer> longestIncreasingSubsequence(int[] arr) {
        return longestIncreasingSubsequence(arr, 0, 0);
    }
    
    private static List<Integer> longestIncreasingSubsequence(int[] arr, int i, int min) {
        if (i == arr.length) {
            return new LinkedList<Integer>();
        }
        
        List<Integer> maxList = longestIncreasingSubsequence(arr, i+1, min);
        
        if (arr[i] > min) {
            List<Integer> includeI = longestIncreasingSubsequence(arr, i+1, arr[i]);
            includeI.add(0, arr[i]);
            if (includeI.size() > maxList.size()) maxList = includeI;
        }
        
        return maxList;
    }
    
    public static void main(String[] args) {
        System.out.println(longestIncreasingSubsequence(new int[]{9,6,1,5,10,13,55,12}));
        System.out.println(longestIncreasingSubsequence(new int[]{9,8,7,6,5,4,3,2,1}));
    }
}

File: python3/LongestIncreasingSubsequence.py
class LongestIncreasingSubsequence:
    def longest_increasing_subsequence(self, l):
        return self.longest_increasing_subsequence_helper(l, 0, 0)

    def longest_increasing_subsequence_helper(self, arr, i, curr_min):
        if i == len(arr):
            return []

        max_list = self.longest_increasing_subsequence_helper(arr, i + 1, curr_min)
        
        if arr[i] > curr_min:
            includeI = self.longest_increasing_subsequence_helper(arr, i + 1, arr[i])
            includeI.insert(0, arr[i])
            if len(includeI) > len(max_list):
                max_list =includeI
        
        return max_list

test = LongestIncreasingSubsequence()
l = [9,6,1,5,3,7,55,12]
print("Result: ",test.longest_increasing_subsequence(l))
